events/src/utils.py

```diff
@@ -313,6 +314,56 @@ def plugin_webhook(message):
     webhooks_pool.send(message["event"], message["ts"], message_json)
 
 
+class NamedTempOpen(object):
+    """
+        This class is used to create a temporary file which is then written to with contents.
+        The temp file is then persisted with the give name by calling os.rename().
+        This class is used to avoid the data loss or truncation in case of multiple processes
+        writing to the same file without the use of fcntl locks.
+
+        The temporary file is created in the dest dir of the file.
+    """
+
+    def __init__(self, filename, open_mode, *args, **kwagrs) -> None:
+        self.filename = filename
+        self.open_mode = open_mode
+        self.open_args = args
+        self.open_kwargs = kwagrs
+        self.working_dir = "."
+        self.fileobj = None
+        self.file_persisted = False
+
+        if os.path.dirname(self.filename) != "":
+            self.working_dir = os.path.dirname(self.filename)
+
+    def persist(self):
+        os.rename(self.fileobj.name, self.filename)
+        self.file_persisted = True
+
+    def __enter__(self):
+        try:
+            tfile = NamedTemporaryFile(mode=self.open_mode,
+                                       delete=False,
+                                       prefix='.',
+                                       dir=self.working_dir,
+                                       *self.open_args,
+                                       **self.open_kwargs)
+
+        except BaseException:
+            raise
+
+        # Monkey patch the persist function onto created temp file object
+        tfile.persist = self.persist
+
+        self.fileobj = tfile
+        return self.fileobj
+
+    def __exit__(self, ex_type, ex_val, ex_tb):
+        self.fileobj.close()
+
+        if not self.file_persisted:
```

Comment:
Instead of that, we can simply check one of the arguments, for example `ex_type`. If it's `None`, then the `with` block completed successfully, so we can rename the file. If it's not `None`, it means that an uncaught exception has been raised. In this case we can delete the file.

Reply:
Got it.